tests: adding test cases for counter RTC on nucleo_g071 or g4xx boards

FRASTM opened the pull request:
This patch enables the RTC as counter feature so that the testcase tests/drivers/counter/counter_basic_api
can run on the nucleo_g071rb board from STMicroelectronics
can run on the nucleo_g474re or nucleo_g431rb board from STMicroelectronics

$ west build -p auto -b nucleo_gxxx tests/drivers/counter/counter_basic_api/

Signed-off-by: Francois Ramu francois.ramu@st.com

FRASTM committed:
drivers: counter: rtc for stm32g0 soc series

This patch set the EXTI line 19 as the RTC
on the stm32g0x serie from STMicroelectronics.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
test timeout

FRASTM commented:
On the stm32g0xx soc, before enabling the RTC clock in the RCC_APBENR1 (STM32_CLOCK_BUS_APB1),
it is required to enable the PWR domain in the same RCC_APBENR1 register. Two bits are set:
-> `clocks = <&rcc STM32_CLOCK_BUS_APB1 0x10000400>;`



- Note that on G4 soc, similar requirement is applied but the reset value of the RCC_APBENR1 is 0x400, meaning that RTC is enabled by default. Consequently for stm32g4.dtsi only has : `clocks = <&rcc STM32_CLOCK_BUS_APB1 0x10000000>;`

FRASTM commented:
rebase

erwango commented on dts/arm/st/g0/stm32g0.dtsi:
```diff
@@ -111,6 +111,16 @@
+			compatible = "st,stm32-rtc";
+			reg = <0x40002800 0x400>;
+			interrupts = <2 0>;
+			clocks = <&rcc STM32_CLOCK_BUS_APB1 0x10000400>;
```
Please enable PWR in clock control's `config_enable_default_clocks` as this is done for other series.
Also applies to G4

FRASTM replied:
according to the ref man of the stm32G0, the access to the backup domain is similar to stm32G4, consequently, the same sequence is expected for both series: 
1. enable the power interface clock
2.  set the DBP bit in the Power Control register 1 (PWR_CR1)

FRASTM committed:
soc: arm: stm32g4 add rtc feature on this serie

This patch enables the rtc clock on the stm32g4 soc
from STMicroelectronics.
Even if the set by default (reset value of theRCC_APB1ENR)
the bit is marked as 1.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
for both stm32g0 and stm32g4 soc:
- select the RCC clock register for enabling RTC in the `STM32_CLOCK_BUS_APB1`
- enable the PWR clock by default in the `config_enable_default_clocks`  configuration at init

erwango commented on soc/arm/st_stm32/stm32g0/soc.h:
```diff
@@ -44,6 +44,12 @@
+#if defined(CONFIG_COUNTER_RTC_STM32)
+#include <stm32g0xx_ll_rtc.h>
+#include <stm32g0xx_ll_exti.h>
+#include <stm32g0xx_ll_pwr.h>
```
Is pwr ll required here (or should be moved to CONFIG_CLOCK_CONTROL_STM32_CUBE) ?

FRASTM replied:
yes, moved to the CONFIG_CLOCK_CONTROL_STM32_CUBE section

FRASTM replied:
yes, moved to the CONFIG_CLOCK_CONTROL_STM32_CUBE section

erwango reviewed (changes requested):
@FRASTM the last PR would need to be reworked as the content doesn't match the description

ABOSTM reviewed (approved):
LGTM

erwango commented:
Moving milestone to V2.5.0 as this is not a bug fix or doc update and now too late to be merged in stabilization period

FRASTM committed:
boards: arm: stm32g4xx add rtc feature on the nucleo board

This patch includes the rtc in the doc for the
nucleo_g474re and nucleo_g431rb boards
from STMicroelectronics

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
rebase

FRASTM committed:
soc: arm: stm32g0 add rtc feature on this serie

This patch enables the rtc clock on the stm32g0 soc
from STMicroelectronics.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: clock_control: stm32g0 soc enables the PWR clock

After system reset, the PWR interface clock must be enabled
by setting the PWREN bit of the RCC_APBENR1
This sequence is needed to use the RTC.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
boards: arm: stm32g071 add rtc feature on the nucleo board

This patch enables the rtc so that the testcase
tests/drivers/counter/counter_basic_api
can run on this nucleo_g071rb board
also when running sanity check

Signed-off-by: Francois Ramu <francois.ramu@st.com>

str4t0m commented on soc/arm/st_stm32/stm32g0/soc.h:
```diff
@@ -33,6 +33,7 @@
 #include <stm32g0xx_ll_utils.h>
 #include <stm32g0xx_ll_bus.h>
 #include <stm32g0xx_ll_rcc.h>
+#include <stm32g0xx_ll_pwr.h>
```
These includes are no longer required since  #30163 was merged.

str4t0m commented on soc/arm/st_stm32/stm32g0/soc.h:
```diff
@@ -47,6 +48,11 @@
 #include <stm32g0xx_ll_i2c.h>
 #endif
 
+#if defined(CONFIG_COUNTER_RTC_STM32)
+#include <stm32g0xx_ll_rtc.h>
+#include <stm32g0xx_ll_exti.h>
+#endif
+
```
Same.

str4t0m reviewed (changes requested):
I have tested this pr on the nucleog071. The changes in soc.h are no longer necessary but otherwise I think this could be rebased an merged.

FRASTM commented:
rebase

MaureenHelm commented on boards/arm/nucleo_g071rb/doc/index.rst:
```diff
@@ -100,6 +100,8 @@ The Zephyr nucleo_g071rb board configuration supports the following hardware fea
 +-----------+------------+-------------------------------------+
 | CLOCK     | on-chip    | reset and clock control             |
 +-----------+------------+-------------------------------------+
+| COUNTER   | on-chip    | rtc                                 |
```
Please remove the conflicts lines in the commit message:
```
boards: arm: stm32g071 add rtc feature on the nucleo board

This patch enables the rtc so that the testcase
tests/drivers/counter/counter_basic_api
can run on this nucleo_g071rb board
also when running sanity check

Signed-off-by: Francois Ramu <francois.ramu@st.com>

# Conflicts:
#	boards/arm/nucleo_g071rb/nucleo_g071rb.yaml
```

MaureenHelm commented on boards/arm/nucleo_g474re/doc/index.rst:
```diff
@@ -119,6 +119,8 @@ The Zephyr nucleo_g474re board configuration supports the following hardware fea
 +-----------+------------+-------------------------------------+
 | ADC       | on-chip    | adc                                 |
 +-----------+------------+-------------------------------------+
+| COUNTER   | on-chip    | rtc                                 |
```
Need to also add counter to the supported list in the board yaml for nucleo_g474re and nucleo_g431rb boards

FRASTM replied:
nucleo_g474re and nucleo_g431rb have -counter in their _yaml_ and rtc in their _dts_
